nemo/collections/nlp/nm/losses/sgd_loss.py

```diff
@@ -0,0 +1,219 @@
+    @add_port_docs
+    def input_ports(self):
+        """Returns definitions of module input ports.
+            logit_intent_status (float): Output of SGD model
+            intent_status_labels (int): Intent labels
+            logit_req_slot_status (float): Output of SGD model
+            requested_slot_status (float): Takes value 1 if the corresponding slot is requested, 0 otherwise
+            req_slot_mask (bool): Masks requested slots not used for the particular service
+            logit_cat_slot_status (float): Output of SGD model
+            categorical_slot_status (int): The status of each categorical slot in the service
+            cat_slot_status_mask (bool): Masks categorical slots not used for the particular service
+            logit_cat_slot_value (float): Output of SGD model
+            categorical_slot_values (int): The index of the correct value for each categorical slot
+            logit_noncat_slot_status (float): Output of SGD model
+            noncategorical_slot_status (int): The status of each noncategorical slot in the service
+            non_cat_slot_status_mask (bool): masks noncategorical slots not used for the particular service
+            logit_noncat_slot_start (float): Output of SGD model
+            logit_noncat_slot_end (float): Output of SGD model
+            noncategorical_slot_value_start (int): The index of the starting subword corresponding to the slot span for a non-categorical slot value
+            noncategorical_slot_value_end (int): The index of the ending (inclusive) subword corresponding to the slot span for a non-categorical slot value
+        """
+        return {
+            "logit_intent_status": NeuralType(('B', 'T', 'C'), LogitsType()),
+            "intent_status_labels": NeuralType(('B'), LabelsType()),
+            "logit_req_slot_status": NeuralType(('B', 'T'), LogitsType()),
+            "requested_slot_status": NeuralType(('B', 'T'), LabelsType()),
+            "req_slot_mask": NeuralType(('B', 'T'), ChannelType()),
+            "logit_cat_slot_status": NeuralType(('B', 'T', 'C'), LogitsType()),
+            "categorical_slot_status": NeuralType(('B', 'T'), LabelsType()),
+            "cat_slot_status_mask": NeuralType(('B'), ChannelType()),
+            "logit_cat_slot_value": NeuralType(('B', 'T', 'C'), LogitsType()),
+            "categorical_slot_values": NeuralType(('B', 'T'), LabelsType()),
+            "logit_noncat_slot_status": NeuralType(('B', 'T', 'C'), LogitsType()),
+            "noncategorical_slot_status": NeuralType(('B', 'T'), LabelsType()),
+            "non_cat_slot_status_mask": NeuralType(('B'), ChannelType()),
+            "logit_noncat_slot_start": NeuralType(('B', 'T', 'C'), LogitsType()),
+            "logit_noncat_slot_end": NeuralType(('B', 'T', 'C'), LogitsType()),
+            "noncategorical_slot_value_start": NeuralType(('B', 'T'), LabelsType()),
+            "noncategorical_slot_value_end": NeuralType(('B', 'T'), LabelsType()),
+        }
```

Comment:
wooow! this is crazy...! :)

Reply:
not sure what will happen with the description of this function, @add_port_docs is making some magic processing down there...